src/serializer/ResidualHeapVisitor.js

```diff
@@ -400,26 +394,66 @@ export class ResidualHeapVisitor {
     }
   }
 
-  visitValueSet(val: ObjectValue): void {
-    let kind = val.getKind();
+  visitValueWeakMap(val: ObjectValue): void {
+    invariant(val.getKind() === "WeakMap");
+    let entries = val.$WeakMapData;
 
-    let entries;
-    if (kind === "Set") {
-      entries = val.$SetData;
-    } else {
-      invariant(kind === "WeakSet");
-      entries = val.$WeakSetData;
-    }
     invariant(entries !== undefined);
     let len = entries.length;
 
+    for (let i = 0; i < len; i++) {
+      this._withUnrelatedScope(this.scope, () => {
+        let entry = entries[i];
+        let key = entry.$Key;
+        let value = entry.$Value;
+
+        if (key !== undefined && value !== undefined) {
+          if (this.values.has(key)) {
+            this.visitValue(key);
+            this.visitValue(value);
+          } else {
+            this._withUnrelatedScope(this.scope, () => {
```

Comment:
This is queue the check one more time. But that's not sufficient for fixpoint computation. It needs to always rerun this logic.